# IBM HR Analytics: Employee Attrition – Data Cleaning

# This notebook loads the dataset, checks data quality, and prepares clean features for analysis.

# 1. Import Libraries

In [1]:
import pandas as pd

# 2. Load Data

In [ ]:
df=pd.read_csv("data\hr-employee-attrition.csv")

# 3. Explore Data

# 3.1 Dataset dimensions

In [3]:
df.shape

(1470, 35)

The dataset contains 1,470 employee records and 35 HR attributes.

# 3.2 Preview first 5 rows

In [4]:
df.head()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,...,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,...,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,...,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,...,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,...,4,80,1,6,3,3,2,2,2,2


The display shows the first 5 employee records to help understand the dataset's structure and content.

# 3.3 Dataset structure and data types

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

The dataset includes 26 numerical features (e.g., age, income, tenure) and 9 categorical features (e.g., department, job role, gender). This mix supports both statistical analysis and HR group comparisons.

# 3.4 Missing values per column

In [6]:
df.isna().sum()

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince

No missing values were found, meaning all employee records are complete and reliable for analysis.

# 3.5 Count duplicate rows

In [7]:
df[df.duplicated()].shape

(0, 35)

No duplicate employee records were found. All 1,470 rows are unique.

# 4. Attrition Distribution

In [8]:
df["Attrition"].value_counts()

Attrition
No     1233
Yes     237
Name: count, dtype: int64

Out of 1,470 employees, 1,233 stayed with the company and 237 left.

# 4.2 Percentage Distribution

In [9]:
df["Attrition"].value_counts(normalize=True)

Attrition
No     0.838776
Yes    0.161224
Name: proportion, dtype: float64

About 16% of employees left the company, confirming that attrition is a meaningful target for predictive modeling.

# 5. Unique Values per Column

In [10]:
df.nunique()

Age                           43
Attrition                      2
BusinessTravel                 3
DailyRate                    886
Department                     3
DistanceFromHome              29
Education                      5
EducationField                 6
EmployeeCount                  1
EmployeeNumber              1470
EnvironmentSatisfaction        4
Gender                         2
HourlyRate                    71
JobInvolvement                 4
JobLevel                       5
JobRole                        9
JobSatisfaction                4
MaritalStatus                  3
MonthlyIncome               1349
MonthlyRate                 1427
NumCompaniesWorked            10
Over18                         1
OverTime                       2
PercentSalaryHike             15
PerformanceRating              2
RelationshipSatisfaction       4
StandardHours                  1
StockOptionLevel               4
TotalWorkingYears             40
TrainingTimesLastYear          7
WorkLifeBa

Columns ‘EmployeeCount’, ‘Over18’, and ‘StandardHours’ each contain only one unique value and do not contribute to the analysis.

# 6. Remove Columns With No Variance

In [11]:
df = df.drop(['EmployeeCount', 'Over18', 'StandardHours'], axis=1)

These fields provide no useful information for understanding attrition.

# 7. Rename Columns to Professional snake_case

# 7.1 Automatic cleaning

In [12]:
df.columns = (
    df.columns
      .str.strip()
      .str.lower()
      .str.replace(" ", "_")
      .str.replace("-", "_")
)

# 7.2 Manual improvements for readability

In [13]:
df = df.rename(columns={
    "businesstravel": "business_travel",
    "dailyrate": "daily_rate",
    "distancefromhome": "distance_from_home",
    "educationfield": "education_field",
    "employeenumber": "employee_number",
    "environmentsatisfaction": "environment_satisfaction",
    "hourlyrate": "hourly_rate",
    "jobinvolvement": "job_involvement",
    "joblevel": "job_level",
    "jobrole": "job_role",
    "jobsatisfaction": "job_satisfaction",
    "maritalstatus": "marital_status",
    "monthlyincome": "monthly_income",
    "monthlyrate": "monthly_rate",
    "numcompaniesworked": "num_companies_worked",
    "percentsalaryhike": "percent_salary_hike",
    "performancerating": "performance_rating",
    "relationshipsatisfaction": "relationship_satisfaction",
    "stockoptionlevel": "stock_option_level",
    "totalworkingyears": "total_working_years",
    "trainingtimeslastyear": "training_times_last_year",
    "worklifebalance": "work_life_balance",
    "yearsatcompany": "years_at_company",
    "yearsincurrentrole": "years_in_current_role",
    "yearssincelastpromotion": "years_since_last_promotion",
    "yearswithcurrmanager": "years_with_curr_manager"
})

Column names are now clean, consistent, and professional for analysis and modeling.

# 8. Convert Attrition to Binary

In [14]:
df['attrition_flag'] = df['attrition'].map({'Yes': 1, 'No': 0})

This numeric attrition variable (1 = left, 0 = stayed) enables predictive modeling.

# 9. Save Clean File

In [15]:
df.to_csv("data/hr_clean.csv", index=False)

The cleaned dataset is saved for use in analysis and modeling.

# 10. Final Summary

# Summary:
# - Loaded 1,470 employee records
# - Verified no missing or duplicate data
# - Removed 3 non-informative columns
# - Renamed all columns to professional snake_case format
# - Created binary attrition_flag for modeling
# - Saved cleaned dataset to data/hr_clean.csv